### PR TITLE
Exposure cube - abritrary ndarrays accepted as input for offset and energy in aeff interp

### PR DESCRIPTION
**Pull request by tibaldo**
@joleroi please review, I implemented this as we discussed yesterday within the interpolation function. I will use this to implement 3D exposure cube, and we can merge once this is working

### CODE REVIEW

**Comment by joleroi** [comment]
In order to make sure everything is working correctly I would be useful to add tests. 
In gammapy/irf/test_effective_area.py there are already some test that test different input shapes. Can you add tests there to also test for 2D input arrays?

**Comment by cdeil** [comment]
My 2 cents on tests:

My experience is that high-level tests are more important than low-level unit tests, so if @tibaldo wants to prefers this now and then add a high-level test that computes an exposure cube and calls this with an array later, I'd say it's OK to merge this without extra test now.

Of course having lots of unit tests is great, but it also slows people down, writing the tests becomes more work than implementing features.

So bottom line: my suggestion for Gammapy would be to encourage / ask contributors to add high-level and unit tests, but it's also OK to say "no, I just want this feature to be merged" or "I'll add tests later".

**Comment by joleroi** [comment]
@cdeil: In principle I agree, but not in this case. An error at this point might be difficult to spot later and is very easy to avoid now, with a 3 line test that was probably made in an IPython notebook or similar anyways. An error at this point might also have implication for several high-level use cases. It also comes with the benefit that another Interpolator that might be added in the future is automatically tested.

Of course I am not keeping anyone from merging their contributions if they want to, it was just a suggestion.

**Comment by tibaldo** [comment]
Exposure cube calculation added. Also Includes a method in SpectralCube class to read a count cube file (that I used for testing). @joleroi @cdeil please review and let me know if it is okay to merge

File: gammapy/irf/exposure.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from ..data import SpectralCube
+from astropy.coordinates import SkyCoord, Angle
+
+import logging
+
+__all__ = [
+    'exposure_cube'
+]
+
+log = logging.getLogger(__name__)
```
Comment (cdeil) [inline gammapy/irf/exposure.py]: remove `import logging` and this line ... it's only needed from files where we do want to log, which is rare and I don't think needs to be added "just in case we want to emit log messages later"

File: gammapy/irf/exposure.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def exposure_cube(pointing,
+                  livetime,
+                  aeff2D,
+                  ref_cube):
+    """Exposure cube
+
+        Parameters
```
Comment (cdeil) [inline gammapy/irf/exposure.py]: Detent this ... it should be at 4 spaces indentation just like the code inside the function.

File: gammapy/irf/exposure.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        ----------
+        pointing : `~astropy.coordinates.SkyCoord`
+            pointing direction
+        livetime : `~astropy.units.Quantity`
+            livetime
+        aeff2D: 'EffectiveAreaTable2D'
+            effective area table
+        ref_cube: 'SpectralCube'
```
Comment (cdeil) [inline gammapy/irf/exposure.py]: Can you try running `python setup.py build_sphinx` and check the HTML API docs locally?
I think you have to reference the class from another file more explicitly for the cross-link to work, no?
And I think there has to be a space before the colon for the formatting to be parsed correctly by Sphinx.
```
       ref_cube : '~gammapy.spectrum.SpectralCube'
```

File: gammapy/data/spectral_cube.py [hunk 4]
```diff
@@ -157,6 +157,30 @@ def read(cls, filename):
...
+        spectral_cube : `SpectralCube`
+            Spectral cube
+        """
+        data = fits.getdata(filename)
+        data = Quantity(data, '')
+        # Note: the energy axis of the FITS cube is unusable.
+        # We only use proj for LON, LAT and do ENERGY ourselves
+        header = fits.getheader(filename)
+        wcs = WCS(header)
+        energy=EnergyBounds.from_ebounds(fits.open(filename)['EBOUNDS'],unit='keV')#keV to match convention in Fermi ST and CTOOLS, TODO read from header
```
Comment (cdeil) [inline gammapy/data/spectral_cube.py]: The comment at the end of the line is too long to be read quickly / on Github / my editor ... can you please put it on a separate line before the code?

**Comment by cdeil** [comment]
I left a few nitpicky comments.

From a quick look, the API and test and code looks good to me.

@joleroi - Do you want to do any review / testing here or should we just merge (once the small issues are adressed) and continue testing / using this from master?

File: gammapy/irf/exposure.py (hunk 1 above)
Comment (tibaldo) [inline gammapy/irf/exposure.py]: ok

**Comment by joleroi** [comment]
No more comments from my side

File: gammapy/irf/exposure.py (hunk 2 above)
Comment (tibaldo) [inline gammapy/irf/exposure.py]: ok

File: gammapy/data/spectral_cube.py (hunk 4 above)
Comment (tibaldo) [inline gammapy/data/spectral_cube.py]: ok

File: gammapy/irf/exposure.py (hunk 3 above)
Comment (tibaldo) [inline gammapy/irf/exposure.py]: Right, was not working properly, fixed now. Thanks!

**Comment by joleroi** [comment]
Don't forget to make an entry in https://github.com/gammapy/gammapy/blob/master/CHANGES.rst :wink:
